### PR TITLE
[DEVMGR] Allow to copy text from Details page of device properties

### PR DESCRIPTION
**Pull request by Oplkill**
In driver properties, in details tab - added popup menu with "copy" element for listview

## Purpose

This will allow to copy driver's information to buffer.

### CODE REVIEW

File: dll/win32/devmgr/properties/advprop.cpp [hunk 1]
```diff
@@ -1897,6 +1897,32 @@ AdvProcDetailsDlgProc(IN HWND hwndDlg,
...
         {
+            case WM_CONTEXTMENU:
+            {
+                if ((HWND)wParam == GetDlgItem(hwndDlg, IDC_DETAILSPROPVALUE))
+                {
+                    WCHAR szColName[255];
+                    INT nSelectedItems = ListView_GetSelectedCount((HWND)wParam);
+                    HMENU hPopup = CreatePopupMenu();
+
+                    if (LoadString(hDllInstance, IDS_COPY, szColName, sizeof(szColName) / sizeof(szColName[0])))
```
Comment (HBelusca) [inline dll/win32/devmgr/properties/advprop.cpp:1908]:
```suggestion
                    if (LoadStringW(hDllInstance, IDS_COPY, szColName, _countof(szColName)))
```

File: dll/win32/devmgr/properties/advprop.cpp [hunk 2]
```diff
@@ -1897,6 +1897,32 @@ AdvProcDetailsDlgProc(IN HWND hwndDlg,
...
+                            EnableMenuItem(hPopup, IDS_MENU_COPY, MF_BYCOMMAND | MF_GRAYED);
+
+                        POINT pt;
+                        GetCursorPos(&pt);
+
+                        TrackPopupMenu(hPopup, TPM_LEFTALIGN, pt.x, pt.y, 0, hwndDlg, NULL);
+                        Ret = TRUE;
+                    }
+
+                    DestroyMenu(hPopup);
```
Comment (HBelusca) [inline dll/win32/devmgr/properties/advprop.cpp:1922]: You create the menu above, but then do the track-menu thing only if the LoadString call succeeded.
Wouldn't it be better to create/track/destroy it  instead _only_ if the LoadString call succeeded?
Alternatively, what would happen instead if you keep the code as it currently is, except that you call the TrackPopupMenu _outside_ of the `if (LoadString(...))` block (and just before calling DestroyMenu)?

File: dll/win32/devmgr/properties/advprop.cpp [hunk 3]
```diff
@@ -1897,6 +1897,32 @@ AdvProcDetailsDlgProc(IN HWND hwndDlg,
...
+                }
```
Comment (HBelusca) [inline dll/win32/devmgr/properties/advprop.cpp:1923]: There is a `break;` missing in there.

File: dll/win32/devmgr/properties/advprop.cpp [hunk 4]
```diff
@@ -1909,6 +1935,51 @@ AdvProcDetailsDlgProc(IN HWND hwndDlg,
...
+                                CloseClipboard();
+                                Ret = TRUE;
+                            }
+                            else
+                            {
+                                GlobalFree(hGlobal);
+                            }
+                        }
+                    }
+                    break;
```
Comment (HBelusca) [inline dll/win32/devmgr/properties/advprop.cpp:1982]:
```suggestion
                        break;
                    }
```

**Review by HBelusca — DISMISSED** [review]
Can you show us a picture how this new menu appears?

**Comment by Oplkill** [comment]
> Can you show us a picture how this new menu appears?

image

Just simple popup menu(or how it calls too - context menu) with only one element - "Copy". That menu will appears on right click mouse button in listview

**Comment by HBelusca** [comment]
> > Can you show us a picture how this new menu appears?
> 
> image
> 
> Just simple popup menu(or how it calls too - context menu) with only one element - "Copy". That menu will appears on right click mouse button in listview

Ah great, OK, so if that menu just have one and only one item "Copy", then I guess, regarding my comment about LoadString + menu creation, that you could create the menu only if the LoadString call succeeds, otherwise not doing anything.

**Comment by Oplkill** [comment]
> > Just simple popup menu(or how it calls too - context menu) with only one element - "Copy". That menu will appears on right click mouse button in listview
> 
> Ah great, OK, so if that menu just have one and only one item "Copy", then I guess, regarding my comment about LoadString + menu creation, that you could create the menu only if the LoadString call succeeds, otherwise not doing anything.

Fixed it

File: dll/win32/devmgr/properties/advprop.cpp [hunk 5]
```diff
@@ -1897,6 +1897,33 @@ AdvProcDetailsDlgProc(IN HWND hwndDlg,
...
+                        INT nSelectedItems = ListView_GetSelectedCount((HWND)wParam);
+                        HMENU hPopup = CreatePopupMenu();
+
+                        AppendMenuW(hPopup, MF_STRING, IDS_MENU_COPY, szColName);
+
+                        if (nSelectedItems <= 0)
+                            EnableMenuItem(hPopup, IDS_MENU_COPY, MF_BYCOMMAND | MF_GRAYED);
+
+                        POINT pt;
+                        GetCursorPos(&pt);
```
Comment (HBelusca) [inline dll/win32/devmgr/properties/advprop.cpp:1917]: Normally the coordinates are passed via the lParam, retrieved via e.g.
```
xPos = GET_X_LPARAM(lParam); 
yPos = GET_Y_LPARAM(lParam); 
```

File: dll/win32/devmgr/properties/advprop.cpp [hunk 6]
```diff
@@ -1897,6 +1897,33 @@ AdvProcDetailsDlgProc(IN HWND hwndDlg,
     {
         switch (uMsg)
         {
+            case WM_CONTEXTMENU:
+            {
+                if ((HWND)wParam == GetDlgItem(hwndDlg, IDC_DETAILSPROPVALUE))
+                {
+                    WCHAR szColName[255];
+
+                    if (LoadStringW(hDllInstance, IDS_COPY, szColName, _countof(szColName)))
```
Comment (HBelusca) [inline dll/win32/devmgr/properties/advprop.cpp:1906]: Note (optional): you could reduce the indentation level by doing:
```c
                    if (!LoadStringW(hDllInstance, IDS_COPY, szColName, _countof(szColName)))
                        break;
```
then continue with the original code as done below.

File: dll/win32/devmgr/properties/advprop.cpp [hunk 7]
```diff
@@ -1909,6 +1936,51 @@ AdvProcDetailsDlgProc(IN HWND hwndDlg,
...
                         }
                         break;
+
+                    case IDS_MENU_COPY:
+                    {
+                        HWND hwndListView = GetDlgItem(hwndDlg, IDC_DETAILSPROPVALUE);
+                        INT nSelectedItems = ListView_GetSelectedCount(hwndListView);
+                        INT nSelectedId = ListView_GetSelectionMark(hwndListView);
+
+                        if (nSelectedId >= 0 && nSelectedItems > 0)
```
Comment (HBelusca) [inline dll/win32/devmgr/properties/advprop.cpp:1946]: Same comment regarding the indentation here too.

File: dll/win32/devmgr/properties/advprop.cpp [hunk 8]
```diff
@@ -1909,6 +1936,51 @@ AdvProcDetailsDlgProc(IN HWND hwndDlg,
...
+                            hGlobal = GlobalAlloc(GHND, MAX_PATH);
+                            pszBuffer = (LPWSTR)GlobalLock(hGlobal);
+
+                            if (pszBuffer == NULL)
+                            {
+                                GlobalFree(hGlobal);
+                                return Ret;
+                            }
+
+                            wsprintfW(pszBuffer, L"%s", szItemName);
```
Comment (HBelusca) [inline dll/win32/devmgr/properties/advprop.cpp:1966]:
```suggestion
                            wsprintf(pszBuffer, L"%s", szItemName);
```
or use a StringSafe function instead.

File: dll/win32/devmgr/properties/advprop.cpp [hunk 9]
```diff
@@ -1909,6 +1936,51 @@ AdvProcDetailsDlgProc(IN HWND hwndDlg,
...
+
+                            ListView_GetItemText(hwndListView,
+                                                 nSelectedId, 0,
+                                                 szItemName,
+                                                 MAX_PATH);
+
+                            hGlobal = GlobalAlloc(GHND, MAX_PATH);
+                            pszBuffer = (LPWSTR)GlobalLock(hGlobal);
+
+                            if (pszBuffer == NULL)
```
Comment (HBelusca) [inline dll/win32/devmgr/properties/advprop.cpp:1960]:
```suggestion
                            if (!pszBuffer)
```

File: dll/win32/devmgr/properties/advprop.cpp [hunk 10]
```diff
@@ -1909,6 +1936,51 @@ AdvProcDetailsDlgProc(IN HWND hwndDlg,
...
+                            TCHAR szItemName[MAX_PATH];
+                            HGLOBAL hGlobal;
+                            LPWSTR pszBuffer;
+
+                            ListView_GetItemText(hwndListView,
+                                                 nSelectedId, 0,
+                                                 szItemName,
+                                                 MAX_PATH);
+
+                            hGlobal = GlobalAlloc(GHND, MAX_PATH);
```
Comment (HBelusca) [inline dll/win32/devmgr/properties/advprop.cpp:1957]:
```suggestion
                            hGlobal = GlobalAlloc(GHND, MAX_PATH);
                            if (!hGlobal)
                                return Ret;
```

**Comment by binarymaster** [comment]
> Just simple popup menu (or how it calls too - context menu) with only one element - "Copy"

On Windows there's a second menu entry "Select all", so the copy operation should be able to work on multiline selection.

**Review by JoachimHenze — APPROVED** [review]
A useful feature. Approved.
It could be made simpler (without requiring any rc changes or LoadString-calls) if we invoked the copy by the pure right-click already, but that would be a bit less self-explanationary, I guess. And people might miss then that we do actually support such a thing.
So I do assume that the cost-benefit-ratio is still ok, in the way that you did.

**Review by HBelusca — APPROVED** [review]

**Comment by HBelusca** [comment]
Oh, by the way @Oplkill , please read that first: https://github.com/reactos/reactos/blob/master/CONTRIBUTING.md#rules-and-recommendations

**Review by binarymaster — CHANGES_REQUESTED** [review]
Hello,
Thanks for your interest in ReactOS!
We ask our contributors [to use their full name when committing](https://github.com/reactos/reactos/blob/master/CONTRIBUTING.md#rules-and-recommendations).
[Please amend your commit with your full name](https://stackoverflow.com/a/3042512), and update this PR.

To change this for future PR's, you can update it for globally (for the system):
```
git config --global user.name "Your Name"
git config --global user.email you@example.com
```

Or just for the current repository:
```
git config user.name "Your Name"
git config user.email you@example.com
```

Also please set up your GitHub account accordingly, so maintainers can use Squash & Merge button without getting the author field modified back to nickname by GitHub:

- Set your full name in [GitHub profile settings](https://github.com/settings/profile) (it should match your `user.name` setting)

Regards,

File: dll/win32/devmgr/lang/de-DE.rc [hunk 11]
```diff
@@ -33,6 +33,7 @@ BEGIN
     IDS_RESOURCE_INTERRUPT "IRQ"
     IDS_RESOURCE_DMA "DMA"
     IDS_RESOURCE_PORT "E/A-Bereich"
+    IDS_COPY "Kopieren"
```
Comment (binarymaster) [inline dll/win32/devmgr/lang/de-DE.rc:36]: Some menu items have mnemonic keys (&) assigned, but some don't.

Comment (binarymaster) [reply dll/win32/devmgr/lang/de-DE.rc:36]: You can copy from notepad's `CMD_COPY`: https://git.reactos.org/?p=reactos.git&a=search&h=HEAD&st=grep&s=CMD_COPY

**Review by JoachimHenze — APPROVED** [review]

**Review by binarymaster — APPROVED** [review]

**Review by HBelusca — APPROVED** [review]

**Comment by ThFabba** [comment]
@HBelusca @binarymaster can you please be a little more diligent with your code review? This should not have been merged.
Brand new code that
* uses TCHARs in a completely broken way
* has an obvious heap buffer overflow
* uses an unsafe string function
is not the standard we want to set.

**Comment by HBelusca** [comment]
I'm going to submit a fix PR for this @ThFabba

**Comment by ThFabba** [comment]
Thanks @HBelusca! Here's what I was going to propose, but I haven't had a chance to test it yet or anything: https://github.com/ThFabba/reactos/commit/e9a8a4b470f58b6ab7d35ef8758e3838b30b1358

**Comment by HBelusca** [comment]
@ThFabba See PR #5790

**Pull request merged**